File: app/src/main/java/com/personal/accountantAssistant/utils/EditableTextsUtils.java
package com.personal.accountantAssistant.utils;

import android.app.Activity;
import android.content.Context;
import android.text.Editable;
import android.text.TextWatcher;
import android.view.View;
import android.view.inputmethod.EditorInfo;
import android.view.inputmethod.InputMethodManager;
import android.widget.EditText;

import com.personal.accountantAssistant.MainActivity;

import java.text.Normalizer;
import java.util.Objects;
import java.util.concurrent.atomic.AtomicReference;

import io.reactivex.functions.Action;

public class EditableTextsUtils {

    static final String EMPTY_VALUE = "";
    static AtomicReference<String> editTextValue = new AtomicReference<>(EMPTY_VALUE);

    public static void initializeListeners(final Activity activity,
                                           final EditText editText,
                                           final Action actionWhenTextChange) {

        editTextValue = new AtomicReference<>(EMPTY_VALUE);

        editText.addTextChangedListener(new TextWatcher() {
            @Override
            public void beforeTextChanged(CharSequence s, int start, int count, int after) {

            }

            @Override
            public void onTextChanged(CharSequence s, int start, int before, int count) {
                setEditTextValue(editText.getText().toString());
                ActionUtils.runAction(actionWhenTextChange);
            }

            @Override
            public void afterTextChanged(Editable s) {
                setEditTextValue(editText.getText().toString());
                ActionUtils.runAction(actionWhenTextChange);
            }
        });

        editText.setOnEditorActionListener((v, actionId, event) -> {
            if (actionId == EditorInfo.IME_ACTION_NEXT ||
                    actionId == EditorInfo.IME_ACTION_GO ||
                    actionId == EditorInfo.IME_ACTION_DONE) {
                hideSoftInputFromWindow(activity, editText);
            }
            return Boolean.TRUE;
        });

        editText.setOnFocusChangeListener((view, hasFocus) -> {
            if (!hasFocus) {
                EditableTextsUtils.hideSoftInputFromWindow(view, editText);
            }
        });
    }

    public static void hideSoftInputFromWindow(final Activity activity,
                                               final EditText editText) {
        final View view = activity.getCurrentFocus();
        if (view != null) {
            hideSoftInputFromWindow(view, editText);
        }
    }

    public static void hideSoftInputFromWindow(final View view,
                                               final EditText editText) {
        if (view != null) {
            final InputMethodManager inputMethodManager = (InputMethodManager) MainActivity.getContext().getSystemService(Context.INPUT_METHOD_SERVICE);
            Objects.requireNonNull(inputMethodManager).hideSoftInputFromWindow(view.getWindowToken(), InputMethodManager.HIDE_NOT_ALWAYS);
            editText.clearFocus();
        }
    }

    private static void setEditTextValue(final String textValue) {
        editTextValue.set(textValue);
    }

    public static String getEditTextValue() {
        return editTextValue.get();
    }

    /**
     * Normalizing string, so it will be possible to search.
     */
    public static String toNormalizedString(final String strValue) {
        final String regexTarget = "[^\\p{ASCII}]";
        return Normalizer
                .normalize(strValue, Normalizer.Form.NFD)
                .replaceAll(regexTarget, Constants.EMPTY_STR);
    }

    public static boolean contains(final String currentStr,
                                   final String filterStr) {
        final String normalizedCurrentStr = EditableTextsUtils
                .toNormalizedString(currentStr)
                .toLowerCase();
        final String normalizedFilterName = EditableTextsUtils
                .toNormalizedString(filterStr)
                .toLowerCase();
        return normalizedCurrentStr.contains(normalizedFilterName);
    }
}

File: app/src/main/java/com/personal/accountantAssistant/utils/EditableTextsUtils.kt
package com.personal.accountantAssistant.utils

import android.app.Activity
import android.content.Context
import android.text.Editable
import android.text.TextWatcher
import android.view.KeyEvent
import android.view.View
import android.view.View.OnFocusChangeListener
import android.view.inputmethod.EditorInfo
import android.view.inputmethod.InputMethodManager
import android.widget.EditText
import android.widget.TextView
import com.personal.accountantAssistant.utils.ActionUtils.runAction
import java.text.Normalizer
import java.util.*
import java.util.concurrent.atomic.AtomicReference

object EditableTextsUtils {
    private const val EMPTY_VALUE = ""
    var editTextValue = AtomicReference(EMPTY_VALUE)
    fun initializeListeners(activity: Activity?, editText: EditText, actionWhenTextChange: () -> Unit) {
        editTextValue = AtomicReference(EMPTY_VALUE)
        editText.addTextChangedListener(object : TextWatcher {
            override fun beforeTextChanged(s: CharSequence, start: Int, count: Int, after: Int) {}
            override fun onTextChanged(s: CharSequence, start: Int, before: Int, count: Int) {
                setEditTextValue(editText.text.toString())
                runAction(actionWhenTextChange)
            }

            override fun afterTextChanged(s: Editable) {
                setEditTextValue(editText.text.toString())
                runAction(actionWhenTextChange)
            }
        })
        editText.setOnEditorActionListener { _: TextView?, actionId: Int, _: KeyEvent? ->
            if (actionId == EditorInfo.IME_ACTION_NEXT ||
                    actionId == EditorInfo.IME_ACTION_GO ||
                    actionId == EditorInfo.IME_ACTION_DONE) {
                activity?.let { hideSoftInputFromWindow(it, editText) }
            }
            java.lang.Boolean.TRUE
        }
        editText.onFocusChangeListener = OnFocusChangeListener { view: View?, hasFocus: Boolean ->
            hasFocus.not().apply { hideSoftInputFromWindow(view, editText) }
        }
    }

    fun hideSoftInputFromWindow(activity: Activity?, editText: EditText) {
        activity?.currentFocus?.let { hideSoftInputFromWindow(it, editText) }
    }

    private fun hideSoftInputFromWindow(view: View?, editText: EditText) {
        view?.let {
            val inputMethodManager = it.context.getSystemService(Context.INPUT_METHOD_SERVICE) as InputMethodManager
            inputMethodManager.hideSoftInputFromWindow(it.windowToken, InputMethodManager.HIDE_NOT_ALWAYS)
            editText.clearFocus()
        }
    }

    private fun setEditTextValue(textValue: String) {
        editTextValue.set(textValue)
    }

    fun getEditTextValue(): String {
        return editTextValue.get()
    }

    /**
     * Normalizing string, so it will be possible to search.
     */
    private fun toNormalizedString(strValue: String?): String {
        val regexTarget = "[^\\p{ASCII}]"
        return Normalizer.normalize(strValue, Normalizer.Form.NFD)
                .replace(regexTarget.toRegex(), Constants.EMPTY_STR)
    }

    @JvmStatic
    fun contains(currentStr: String?, filterStr: String?): Boolean {
        val normalizedCurrentStr = toNormalizedString(currentStr).toLowerCase(Locale.ROOT)
        val normalizedFilterName = toNormalizedString(filterStr).toLowerCase(Locale.ROOT)
        return normalizedCurrentStr.contains(normalizedFilterName)
    }
}